src/common/runtime/helper/test_worker.ts

```diff
@@ -2,95 +2,125 @@ import { LogMessageWithStack } from '../../internal/logging/log_message.js';
   ): Promise<void> {
-    this.worker.postMessage({
-      query,
-      expectations,
-      ctsOptions: this.ctsOptions,
-    });
-    const workerResult = await new Promise<LiveTestCaseResult>(resolve => {
-      this.resolvers.set(query, resolve);
-    });
-    rec.injectResult(workerResult);
+    const request: WorkerTestRunRequest = { rec, query, expectations };
+    await this.makeRequestAndRecordResult(this.worker, request);
   }
 }
 
 export class TestWorker extends TestDedicatedWorker {}
 
-export class TestSharedWorker {
-  private readonly ctsOptions: CTSOptions;
+export class TestSharedWorker extends TestBaseWorker {
   private readonly port: MessagePort;
-  private readonly resolvers = new Map<string, (result: LiveTestCaseResult) => void>();
 
   constructor(ctsOptions?: CTSOptions) {
-    this.ctsOptions = { ...(ctsOptions || kDefaultCTSOptions), ...{ worker: 'shared' } };
+    super('shared', ctsOptions);
     const selfPath = import.meta.url;
     const selfPathDir = selfPath.substring(0, selfPath.lastIndexOf('/'));
     const workerPath = selfPathDir + '/test_worker-worker.js';
     const worker = new SharedWorker(workerPath, { type: 'module' });
     this.port = worker.port;
     this.port.start();
-    this.port.onmessage = ev => {
-      const query: string = ev.data.query;
-      const result: TransferredTestCaseResult = ev.data.result;
-      if (result.logs) {
-        for (const l of result.logs) {
-          Object.setPrototypeOf(l, LogMessageWithStack.prototype);
-        }
-      }
-      this.resolvers.get(query)!(result as LiveTestCaseResult);
+    this.port.onmessage = ev => this.onmessage(ev);
+  }
+
+  async run(
+    rec: TestCaseRecorder,
+    query: string,
+    expectations: TestQueryWithExpectation[] = []
+  ): Promise<void> {
+    const request: WorkerTestRunRequest = { rec, query, expectations };
+    await this.makeRequestAndRecordResult(this.port, request);
+  }
+}
 
-      // MAINTENANCE_TODO(kainino0x): update the Logger with this result (or don't have a logger and
-      // update the entire results JSON somehow at some point).
-    };
+export class TestServiceWorker extends TestBaseWorker {
+  constructor(ctsOptions?: CTSOptions) {
+    super('service', ctsOptions);
   }
 
   async run(
     rec: TestCaseRecorder,
     query: string,
     expectations: TestQueryWithExpectation[] = []
   ): Promise<void> {
-    this.port.postMessage({
-      query,
-      expectations,
-      ctsOptions: this.ctsOptions,
-    });
-    const workerResult = await new Promise<LiveTestCaseResult>(resolve => {
-      this.resolvers.set(query, resolve);
+    const [suite, name] = query.split(':', 2);
+    const fileName = name.split(',').join('/');
+    const serviceWorkerPath = `/out/${suite}/webworker/${fileName}.worker.js`;
+
+    const registration = await navigator.serviceWorker.register(serviceWorkerPath, {
+      type: 'module',
     });
-    rec.injectResult(workerResult);
+    await registration.update();
```

Comment:
@beaufortfrancois Is this supposed to be the right way to wait for a registration to be active? It appears to be redundant, and it's really slow (90-100ms), but I can't find any reliable way to wait until the service worker is active. `navigator.serviceWorker.ready` is global, and doesn't work anyway (probably because of the scope). 

The best idea I have so far is
```js
      while (!registration.active) {
        await new Promise(resolve => setTimeout(resolve, 0));
      }
```
which is much faster (5-6ms).

Reply:
`await registration.update();` is actually required to make sure the service worker is registered with the new code.

Steps to reproduce issue:
1. Go to http://localhost:8080/standalone/?worker=service&q=webgpu:api,operation,*
2. Run "webgpu:api,operation,adapter,requestAdapter:*" subtree
3. Run "webgpu:api,operation,adapter,requestAdapterInfo:*" subtree

Without `registration.update();`, the second service worker is the one from the first subtree.

Reply:
LGTM ;)
Please merge